server/sonar-server/src/test/java/org/sonar/server/projectlink/ws/SearchActionTest.java

```diff
@@ -0,0 +1,266 @@
+  public void response_fields() throws IOException {
+    ComponentDto project = insertProject();
+    ComponentLinkDto homepageLink = insertHomepageLink(project.uuid());
+    ComponentLinkDto customLink = insertCustomLink(project.uuid());
+
+    SearchWsResponse response = callByKey(project.key());
+
+    assertThat(response.getLinksCount()).isEqualTo(2);
+    assertThat(response.getLinksList()).extracting(Link::getId, Link::getName, Link::getType, Link::getUrl)
+      .containsOnlyOnce(
+        tuple(homepageLink.getIdAsString(), homepageLink.getName(), homepageLink.getType(), homepageLink.getHref()),
+        tuple(customLink.getIdAsString(), customLink.getName(), customLink.getType(), customLink.getHref()));
+  }
+
+  @Test
+  public void several_projects() throws IOException {
+    ComponentDto project1 = insertProject();
+    ComponentDto project2 = insertProject("another", "abcd");
+    ComponentLinkDto customLink1 = insertCustomLink(project1.uuid());
+    insertCustomLink(project2.uuid());
+
+    SearchWsResponse response = callByKey(project1.key());
+
+    assertThat(response.getLinksCount()).isEqualTo(1);
+    assertThat(response.getLinks(0).getId()).isEqualTo(customLink1.getIdAsString());
+  }
+
+  @Test
+  public void global_admin() throws IOException {
+    userSession.login("login").setGlobalPermissions(SYSTEM_ADMIN);
+    checkItWorks();
+  }
+
+  @Test
+  public void project_admin() throws IOException {
+    userSession.login("login");
+    ComponentDto project = insertProject();
+    userSession.addProjectUuidPermissions(UserRole.ADMIN, project.uuid());
+    checkItWorks(project);
+  }
+
+  @Test
+  public void project_user() throws IOException {
+    userSession.login("login");
+    ComponentDto project = insertProject();
+    userSession.addProjectUuidPermissions(UserRole.USER, project.uuid());
+    checkItWorks(project);
+  }
+
+  @Test
+  public void fail_when_no_project() throws IOException {
+    expectedException.expect(NotFoundException.class);
+    callByKey("unknown");
+  }
+
+  @Test
+  public void fail_if_insufficient_privileges() throws IOException {
+    userSession.anonymous();
+    insertProject();
+
+    expectedException.expect(ForbiddenException.class);
+    callByKey(PROJECT_KEY);
+  }
+
+  @Test
+  public void fail_when_both_id_and_key_are_provided() {
+    insertProject();
+
+    expectedException.expect(IllegalArgumentException.class);
+    ws.newRequest()
+      .setParam(PARAM_PROJECT_KEY, PROJECT_KEY)
+      .setParam(PARAM_PROJECT_ID, PROJECT_UUID)
+      .execute();
+  }
+
+  @Test
+  public void fail_when_no_id_nd_key_are_provided() {
+    insertProject();
+
+    expectedException.expect(IllegalArgumentException.class);
+    ws.newRequest()
+      .setParam(PARAM_PROJECT_KEY, PROJECT_KEY)
+      .setParam(PARAM_PROJECT_ID, PROJECT_UUID)
+      .execute();
+  }
+
+  private ComponentDto insertProject(String projectKey, String projectUuid) {
+    return componentDb.insertComponent(new ComponentDto()
```

Comment:
use `ComponentDbTesting.createProjectDto` so no need to handle uuid path and root uuid normally